### ISSUE TITLE
Allow customization of data delivery behavior of twisted.protocols.loopback.loopbackAsync

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3820]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3820|
|Type|enhancement|
|Created|2009-05-04 18:14:01Z|
|Branch|https://github.com/twisted/twisted/tree/loopback-chunking-parameter-3820|

`loopbackAsync` connects two protocol instances to each other, providing two in-memory transports which send bytes written to them by one protocol object to the `dataReceived` method of the other protocol (much like an actual transport would).

The currently implemented behavior is a little unrealistic.  When the data delivery phase comes, each string written to one transport is delivered to the other protocol separately.  So if a protocol writes "foo", then "bar", then "baz" to its transport, the other protocol will receive "foo" in one call to `dataReceived`, "bar" in a second, and finally "baz" in a third.  A real transport would probably clump these together sometimes, or split them up, or clump them and then split them along different boundaries differently.

Protocols are supposed to deal with this kind of clumping and splitting, and while the behavior implemented now isn't _wrong_ - but it might trick developers into thinking that their code works, when really it only works for the exact chunks their tests produce.

It would be pretty neat to have a thing which could generate several tests automatically, based on a single test method implementation, which exercised lots of different chunking behaviors.

However, for a start, it would be nice if the chunking behavior were a parameter to loopbackAsync, so that tests that _know_ they want to exercise a different code path can.

In particular, I want this for [#10385](https://github.com/twisted/twisted/issues/10385) so that I can exercise the case of multiple SSL messages
being delivered in a single call to `dataReceived`.

<details><summary>Searchable metadata</summary>

```
trac-id__3820 3820
type__enhancement enhancement
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_loopback_chunking_parameter_3820 branches/loopback-chunking-parameter-3820
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests tests
time__1241460841000000 1241460841000000
changetime__1241563052000000 1241563052000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [26836]) Branching to 'loopback-chunking-parameter-3820'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/loopback-chunking-parameter-3820) look good.  Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Fixing a couple thinkos in ticket summary/description

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

OK great, only cosmetics:

 * Please update copyright statements
 * Please try to fit some of the code in loopback in 80 columns (what you can easily do).

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [26854]) Update copyright dates

refs [#3820](https://github.com/twisted/twisted/issues/3820)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [26856]) Shorten long lines

refs [#3820](https://github.com/twisted/twisted/issues/3820)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [26857]) Merge loopback-chunking-parameter-3820

Author: exarkun
Reviewer: therve
Fixes: [#3820](https://github.com/twisted/twisted/issues/3820)

Add an optional parameter to `twisted.protocols.loopback.loopbackAsync`
which allows the caller to control exactly how bytes written to the
loopback transport are delivered to the associated protocol.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
